### ISSUE TITLE
Guassian Blur(Win7 aeroglass, Mac/iOS UI blur) Feature

### ISSUE THREAD

**Issue by sekkit** [vczh-libraries/GacUI#29]
Built-in blur effect would be cool.
With shader that can be achieved easily in Game Engines.  But ordinary desktop's GUI depends on OS implementation.
Is it possible to implement high performance - high quality Blur effect just can be easily enabled by a xml tag?

**Comment by vczh** [comment]
I have not looked into that yet. The default engine is Direct2D, and everything is writing to a render target. If Direct2D enable to blur something that has already been rendered to a render target, then your idea is doable. Do you have any idea?

**Comment by sekkit** [comment]
How do you achieve crossplatform with D2D? Maybe you should reference NoesisGUI or QT because it supports various render backend as well as Data-binding like WPF.
I think if u implement the UI lib with a set of stable backend APIs then it's possible to be crossplatform with the same codebase.
As for blur, I think if you can get the render buffer and apply blur algorithm onto that bitmap, then it should be done. Shader might be a good way to go.
I don't know the difference between NoesisGUI and GacUI, sorry for that because I am only familiar with limited Game GUIs.

**Comment by vczh** [comment]
@sekkit if you take a look at iGac repo you will find out that I achieve cross platform by writing different renderers and windowing provider for each platform I want to support. But unfortunately the maintainer of iGac are not working so the progress doesn't catch up. I get some idea from WPF actually.

So for the blur feature, I need to solve it for each platform if I want to add it.

GacUI are going to be able to switch between themes by calling one function but it is not done yet. The good news is, all fundamental building blocks are done, I just need to write a recursive function to tell the control tree to switch the theme. The bad news is that, I didn't create those 2k xp and win7 themes. Maybe someday I will do that, or maybe users will create by themselves.

**Comment by vczh** [comment]
@sekkit you can track my progress by [reading this file](https://github.com/vczh-libraries/Release/blob/master/TODO.md#before-10)

**Comment by sekkit** [comment]
OS independent, 就是跨平台的aeroglass

**Comment by sekkit** [comment]
FlashGet 3可以一键切Xp主题(内置的)，很漂亮。
Windows的主题不是向后兼容的，这很坑爹。OS决定了用户的审美。
QT要实现这个，也得自己动手。
类似2k, Xp，Win7的主题，如果能跨平台，随便切换，磨砂玻璃也可以随便开启的话，再加一个主题商店，大家自己上传（最好只用改xml，不用做代码，这样UE也能自己做主题了），就比其它UI库要方便的多。
说实在的，Win的主题除了Win8,10，其它都好看。Win10之后微软的GUI已经没有审美了。
RibbonTab，我觉得应该是个主题+layout布局（Tab控件），能应用上面的各种主题而不是office like only。
NeosisGUI，这个也是xaml的，你要不也扩展下游戏领域~~

**Comment by vczh** [comment]
@sekkit 皮肤可以慢慢做，Windows游戏开发的话，Direct2D跟Direct3D10以上都是兼容的，算燃气可以直接用，换个window provider就好了。aeroglass肯定还是尽量寻求操作系统支持，据说Win8之后已经没有这个功能了。

**Comment by sekkit** [comment]
Nope. http://doc.qt.io/qt-5/graphicaleffects.html GUI直接有。自己实现的效率不一定好~
以前试过WPF在非Win7上实现磨砂效果，性能很差，这些消耗高的算法应该做在c++层吧

**Comment by vczh** [comment]
得用D3D写shader做，不能直接CPU坐的

**Comment by vczh** [comment]
https://docs.microsoft.com/en-us/archive/msdn-magazine/2014/june/windows-with-c-high-performance-window-layering-using-the-windows-composition-engine 可能有用

**Comment by roodkcab** [comment]
maybe should use stackblur / boxblus instead of guassian blur
